File: backend/vision/anthropic_provider.py
"""Vision provider backed by the Anthropic Messages API. Reads
ANTHROPIC_API_KEY from the environment -- never hardcode a key."""

from __future__ import annotations

import base64
import os

import anthropic

from backend.vision.base import MealAnalysis, VisionProvider, VisionProviderError

DEFAULT_ANTHROPIC_MODEL = "claude-opus-5"


def _detect_media_type(image_bytes: bytes) -> str:
    if image_bytes.startswith(b"\xff\xd8"):
        return "image/jpeg"
    if image_bytes.startswith(b"\x89PNG\r\n\x1a\n"):
        return "image/png"
    if image_bytes.startswith(b"GIF87a") or image_bytes.startswith(b"GIF89a"):
        return "image/gif"
    if image_bytes[:4] == b"RIFF" and image_bytes[8:12] == b"WEBP":
        return "image/webp"
    return "image/jpeg"


class AnthropicProvider(VisionProvider):
    def __init__(self, model: str | None = None, client: "anthropic.Anthropic | None" = None):
        if client is None and not os.environ.get("ANTHROPIC_API_KEY"):
            raise VisionProviderError(
                "ANTHROPIC_API_KEY is not set. Set it in the environment to use AnthropicProvider."
            )
        self.model = model or os.environ.get("ANTHROPIC_VISION_MODEL", DEFAULT_ANTHROPIC_MODEL)
        self.client = client if client is not None else anthropic.Anthropic()

    def analyze(self, image_bytes: bytes, hint: str | None = None) -> MealAnalysis:
        media_type = _detect_media_type(image_bytes)
        image_b64 = base64.standard_b64encode(image_bytes).decode("utf-8")
        return self._analyze_with_retry(
            lambda prompt: self._call_anthropic(prompt, image_b64, media_type), hint
        )

    def _call_anthropic(self, prompt: str, image_b64: str, media_type: str) -> str:
        try:
            response = self.client.messages.create(
                model=self.model,
                max_tokens=2048,
                output_config={"effort": "low"},
                messages=[
                    {
                        "role": "user",
                        "content": [
                            {
                                "type": "image",
                                "source": {"type": "base64", "media_type": media_type, "data": image_b64},
                            },
                            {"type": "text", "text": prompt},
                        ],
                    }
                ],
            )
        except anthropic.APIError as e:
            raise VisionProviderError(f"Anthropic API call failed: {e}") from e
        return "".join(block.text for block in response.content if block.type == "text")

File: lib/vision/anthropic-provider.ts
import { blobToBase64, resizeAndCompressImage } from "./image";
import { buildMealAnalysisPrompt } from "./prompt";
import { parseMealAnalysis, VisionAnalysisError, type MealAnalysis, type VisionProvider } from "./types";

const MESSAGES_URL = "https://api.anthropic.com/v1/messages";
const ANTHROPIC_VERSION = "2023-06-01";

/** Balances vision quality against per-photo cost for a consumer app users pay for with their own key. */
const DEFAULT_MODEL = "claude-sonnet-5";
/** Cheapest current model, used only for the Settings "Test key" ping. */
const TEST_MODEL = "claude-haiku-4-5";

interface AnthropicContentBlock {
  type: string;
  text?: string;
}

interface AnthropicMessageResponse {
  content: AnthropicContentBlock[];
}

interface AnthropicErrorResponse {
  error?: { type?: string; message?: string };
}

/**
 * Calls the Anthropic Messages API directly from the browser — the photo
 * goes straight from the device to Anthropic and never passes through a
 * Vercel function. `anthropic-dangerous-direct-browser-access` is what
 * makes Anthropic allow a browser-origin CORS request at all.
 */
export class AnthropicBrowserProvider implements VisionProvider {
  readonly id = "anthropic";
  readonly label = "Claude (your API key)";

  constructor(
    private readonly apiKey: string,
    private readonly model: string = DEFAULT_MODEL,
  ) {}

  async analyze(imageBlob: Blob, hint?: string): Promise<MealAnalysis> {
    const compressed = await resizeAndCompressImage(imageBlob);
    const base64 = await blobToBase64(compressed);
    const prompt = buildMealAnalysisPrompt(hint);

    let lastError: unknown;
    for (let attempt = 0; attempt < 2; attempt++) {
      const text = await this.requestText(base64, prompt, attempt);
      try {
        return parseMealAnalysis(text);
      } catch (err) {
        lastError = err;
      }
    }
    throw new VisionAnalysisError(
      "Claude didn't return a valid meal analysis after two attempts. Falling back to manual entry.",
      lastError,
    );
  }

  /** Makes one minimal, cheap request purely to validate the key. */
  async testKey(): Promise<{ ok: true } | { ok: false; error: string }> {
    try {
      await this.callMessages({
        model: TEST_MODEL,
        max_tokens: 8,
        messages: [{ role: "user", content: "Reply with the single word: ok" }],
      });
      return { ok: true };
    } catch (err) {
      return { ok: false, error: err instanceof Error ? err.message : String(err) };
    }
  }

  private async requestText(base64Image: string, prompt: string, attempt: number): Promise<string> {
    const promptText =
      attempt === 0
        ? prompt
        : `${prompt}\n\nYour previous response could not be parsed as valid JSON matching this schema. Return ONLY the raw JSON object, nothing else.`;

    const response = await this.callMessages({
      model: this.model,
      max_tokens: 1024,
      output_config: { effort: "low" },
      messages: [
        {
          role: "user",
          content: [
            { type: "image", source: { type: "base64", media_type: "image/jpeg", data: base64Image } },
            { type: "text", text: promptText },
          ],
        },
      ],
    });

    const textBlock = response.content.find((block) => block.type === "text");
    if (!textBlock?.text) {
      throw new VisionAnalysisError("Claude's response contained no text content.");
    }
    return textBlock.text;
  }

  private async callMessages(body: Record<string, unknown>): Promise<AnthropicMessageResponse> {
    const res = await fetch(MESSAGES_URL, {
      method: "POST",
      headers: {
        "x-api-key": this.apiKey,
        "anthropic-version": ANTHROPIC_VERSION,
        "content-type": "application/json",
        "anthropic-dangerous-direct-browser-access": "true",
      },
      body: JSON.stringify(body),
    });

    if (!res.ok) {
      let message = `Anthropic API request failed with status ${res.status}.`;
      try {
        const errBody = (await res.json()) as AnthropicErrorResponse;
        if (errBody.error?.message) message = errBody.error.message;
      } catch {
        // response body wasn't JSON; keep the generic message
      }
      throw new VisionAnalysisError(message);
    }

    return (await res.json()) as AnthropicMessageResponse;
  }
}
